File: src/main/python/run_whois.py
#!/usr/bin/env python

import sys
import psycopg2
import time
import traceback
import pythonwhois

import psycopg2.extensions

conn = psycopg2.connect(host = sys.argv[1], user = sys.argv[2], password = sys.argv[3], database = sys.argv[4])
conn.set_client_encoding('UTF-8')
cur = conn.cursor()
SLEEP_TIME = 1.5

for i in range(1000):
    cur.execute("""
        lock table domains in SHARE ROW EXCLUSIVE mode;
        UPDATE domains
        SET    started = 'now()'
        WHERE  domain = (SELECT domain FROM domains WHERE started IS NULL LIMIT 1)
        RETURNING domain
    """)
    conn.commit()
    domain = cur.fetchone()[0]
    print('processing %s' % domain)
    try:
        records = pythonwhois.net.get_whois_raw(domain)
        result = ('\n\n' + '=+' * 40 + '\n\n').join(records)
        cur.execute("""
            UPDATE domains
            SET completed = 'now()', status = 'C', message = %s
            WHERE domain = %s
        """, (result, domain))
    except:
        traceback.print_exc()
        cur.execute("""
            UPDATE domains
            SET completed = 'now()', status = 'E', error = %s
            WHERE domain = %s
        """, (traceback.format_exc(), domain))

    conn.commit()

    time.sleep(SLEEP_TIME)

File: src/main/ruby/run_whois.rb
#!/usr/bin/env ruby
#
# Install using: gem install pg whois

Encoding.default_internal = Encoding::UTF_8
Encoding.default_external = Encoding::UTF_8

require 'whois'
require 'pg'

conn = PG::Connection.open(:host => ARGV[0], :user => ARGV[1], :password => ARGV[2], :dbname => ARGV[3])

for i in 0 .. 750
    domain = ''
    conn.transaction do
      res = conn.exec_params('
                lock table domains in SHARE ROW EXCLUSIVE mode;
                UPDATE domains
                SET    started = \'now()\'
                WHERE  domain = (SELECT domain FROM domains WHERE started IS NULL LIMIT 1)
                RETURNING domain')
      domain = res.getvalue(0, 0)
    end
    begin
        z
        $stderr.puts "looking up #{domain}"
        r = Whois.whois(domain)
        sv = r.server.class.name.demodulize
        res = r.parts.map(&:body).join("\n\n" + ("=+" * 40) + "\n\n")
        conn.transaction do
          conn.exec_params('
              UPDATE domains
              SET completed = \'now()\', status = \'C\', server = $1, message = $2
              WHERE domain = $3
           ', [sv, res, domain])
        end
    rescue Exception => e
        $stderr.puts e
        conn.transaction do
          conn.exec_params('
              UPDATE domains
              SET completed = \'now()\', status = \'E\', error = $1
              WHERE domain = $2
           ', [e, domain])
        end
    end
    sleep(3)
end



